### PR TITLE
[PassManager] Add pretty stack frames

### PR DESCRIPTION
**Pull request by nikic**
In NewPM pass managers, add a "pretty stack frame" that tells you which pass crashed while running which function.

For example `opt -O3 -passes-ep-peephole=trigger-crash-function test.ll` will print something like this:

```
Stack dump:
0.	Program arguments: build/bin/opt -S -O3 -passes-ep-peephole=trigger-crash-function test.ll
1.	Running pass "function<eager-inv>(mem2reg,instcombine<max-iterations=1;no-use-loop-info;no-verify-fixpoint>,trigger-crash-function,simplifycfg<bonus-inst-threshold=1;no-forward-switch-cond;switch-range-to-icmp;no-switch-to-lookup;keep-loops;no-hoist-common-insts;no-sink-common-insts;speculate-blocks;simplify-cond-branch>)" on module "test.ll"
2.	Running pass "trigger-crash-function" on function "fshl_concat_i8_i8"
```

While the crashing pass is usually evident from the stack trace, this also shows which function triggered the crash, as well as the pipeline string for the pass (including options).

Similar functionality existed in the LegacyPM.

-----

I'm not sure whether the way I did this is the best way to do it -- took me a while to figure out the correct combination of C++ template magic to make this work.

### CODE REVIEW

**Comment by fhahn** [comment]
I originally tried to improve this in 128745cc2681c284bc6d0150a319673a6d6e8424 which did things a bit more complicated fashion. That approach turned out to be relatively expensive and was reverted and I never had time to get back to it. Would be good to check the compile-time impact just to be sure

**Comment by nikic** [comment]
> I originally tried to improve this in [128745c](https://github.com/llvm/llvm-project/commit/128745cc2681c284bc6d0150a319673a6d6e8424) which did things a bit more complicated fashion. That approach turned out to be relatively expensive and was reverted and I never had time to get back to it. Would be good to check the compile-time impact just to be sure

Looks like there is close to no impact: https://llvm-compile-time-tracker.com/compare.php?from=60984f5be90d3f0378d739792adbd4c5e3dcc0e2&to=3d48d4932284fd29d0ede70f0ea6834a7c204d5e&stat=instructions:u

**Comment by nikic** [comment]
> I would have liked to print the pipeline string instead of pass name (so we also get pass options), but the necessary information for that doesn't seem to be available.

I just realized that this *is* available via PI -> PIC, just currently not exposed in the API. I think I'll change this PR to make use of that.

**Comment by aeubanks** [comment]
> > I would have liked to print the pipeline string instead of pass name (so we also get pass options), but the necessary information for that doesn't seem to be available.
> 
> I just realized that this _is_ available via PI -> PIC, just currently not exposed in the API. I think I'll change this PR to make use of that.

It's only available when `shouldPopulateClassToPassNames()` is true, unsure of the compile time impact of always populating the pass name map

**Comment by nikic** [comment]
> > I just realized that this _is_ available via PI -> PIC, just currently not exposed in the API. I think I'll change this PR to make use of that.
> 
> It's only available when `shouldPopulateClassToPassNames()` is true, unsure of the compile time impact of always populating the pass name map

Always initializing the map does have measurable overhead: http://llvm-compile-time-tracker.com/compare.php?from=36c6632eb43bf67e19c8a6a21981cf66e06389b4&to=22e3ed4c5dfd93dcc3684410515ee053ace2edf3&stat=instructions:u

I've put up https://github.com/llvm/llvm-project/pull/96321 to use lazy initialization instead.

**Comment by nikic** [comment]
Okay, I think this is ready for review now.

File: llvm/lib/IR/PassManager.cpp [hunk 1]
```diff
@@ -13,6 +13,7 @@
 using namespace llvm;
 
 namespace llvm {
+
```
Comment (fhahn) [inline llvm/lib/IR/PassManager.cpp:16]: unrelated change?

**Review by fhahn — APPROVED** [review]
LGTM assuming compile-time impact is fine, thanks!

File: llvm/include/llvm/IR/PassManager.h [hunk 2]
```diff
@@ -224,11 +224,21 @@ class PassManager : public PassInfoMixin<
   std::vector<std::unique_ptr<PassConceptT>> Passes;
 };
 
+template <typename IRUnitT>
+void printIRUnitNameForStackTrace(raw_ostream &OS, const IRUnitT &IR);
```
Comment (aeubanks) [inline llvm/include/llvm/IR/PassManager.h:228]: can we put these in PassManagerImpl.h?

**Review by aeubanks — APPROVED** [review]

File: llvm/include/llvm/IR/PassManager.h (hunk 2 above)
Comment (nikic) [reply llvm/include/llvm/IR/PassManager.h:228]: In theory, no, because other places using PassManagerImpl.h also need to specialize these for other types. In practice, yes, we could, because all non-Module/Function users actually specialize the PassManager::run implementation (and as such currently aren't part of the stack trace).

Architecturally, I think putting the in PassManager.h is correct, though possibly this whole implementation is unnecessarily general given that we only end up using it for Module/Function right now.
